hosts/jenkins-controller/configuration.nix

```diff
@@ -0,0 +1,179 @@
+    before = ["nix-daemon.service"];
+    requires = ["network.target"];
+    wantedBy = [
+      # nix-daemon is socket-activated, and having it here should be sufficient
+      # to fetch the keys whenever a jenkins job connects to the daemon first.
+      # This means this service will effectively get socket-activated on the
+      # first nix-daemon connection.
+      "nix-daemon.service"
+    ];
+    serviceConfig = {
+      Type = "oneshot";
+      RemainAfterExit = true;
+      EnvironmentFile = "/var/lib/fetch-build-ssh-key/env";
+      Restart = "on-failure";
+    };
+    script = let
+      get-secret = pkgs.writers.writePython3 "get-secret" {
+        libraries = with pkgs.python3.pkgs; [azure-keyvault-secrets azure-identity];
+      } (builtins.readFile ./get_secret.py);
+    in ''
+      umask 077
+      mkdir -p /etc/secrets/
+      ${get-secret} > /etc/secrets/remote-build-ssh-key
+    '';
+  };
+
+  # populate-known-hosts populates /root/.ssh/known_hosts with all hosts in the
+  # builder subnet.
+  systemd.services.populate-known-hosts = {
+    after = ["network.target"];
+    before = ["nix-daemon.service"];
+    requires = ["network.target"];
+    wantedBy = [
+      # nix-daemon is socket-activated, and having it here should be sufficient
+      # to fetch the keys whenever a jenkins job connects to the daemon first.
+      # This means this service will effectively get socket-activated on the
+      # first nix-daemon connection.
+      "nix-daemon.service"
+    ];
+    serviceConfig = {
+      Type = "oneshot";
+      RemainAfterExit = true;
+      Restart = "on-failure";
+    };
+    script = ''
+      umask 077
+      mkdir -p /root/.ssh
+      ${pkgs.openssh}/bin/ssh-keyscan -f /var/lib/builder-keyscan/scanlist -v -t ed25519 > /root/.ssh/known_hosts
+    '';
+  };
+
+  # Define a fetch-binary-cache-signing-key unit populating
+  # /etc/secrets/nix-signing-key from Azure Key Vault.
+  # Make it before and requiredBy nix-daemon.service.
+  systemd.services.fetch-binary-cache-signing-key = {
+    after = ["network.target"];
+    before = ["nix-daemon.service"];
+    requires = ["network.target"];
+    wantedBy = [
+      # nix-daemon is socket-activated, and having it here should be sufficient
+      # to fetch the keys whenever a jenkins job connects to the daemon first.
+      # This means this service will effectively get socket-activated on the
+      # first nix-daemon connection.
+      "nix-daemon.service"
+    ];
+    serviceConfig = {
+      Type = "oneshot";
+      RemainAfterExit = true;
+      EnvironmentFile = "/var/lib/fetch-binary-cache-signing-key/env";
+      Restart = "on-failure";
+    };
+    script = let
+      get-secret = pkgs.writers.writePython3 "get-secret" {
+        libraries = with pkgs.python3.pkgs; [azure-keyvault-secrets azure-identity];
+      } (builtins.readFile ./get_secret.py);
+    in ''
+      umask 077
+      mkdir -p /etc/secrets/
+      ${get-secret} > /etc/secrets/nix-signing-key
```

Comment:
Nix is pretty bad when it comes to signing, and only supports the private key being present on some filesystem - here's the equivalent desk in the existing Hydra setup: https://github.com/tiiuae/ghaf-infra/blob/7f88348311e3dca947e4f40219968c1069e420d5/services/hydra/default.nix#L120

I agree signing shouldn't happen on the box, but instead we should offload the signing to an HSM/Key vault, so even the jenkins-controller can't see the secret. As long as that other signer can establish machine identity and supports signing arbitrary ed25519 payloads (the NARInfo fingerprint), the rest of Nix doesn't care how the signature was made.

I already wrote most of the building blocks w.r.t calculating fingerprints, adding signatures to NARInfo files etc. I'll add a futurework item for this in the docs.

Reply:
As discussed, in a short term, the signing can be simply replaced with a API call to the Keyvault. The keyvault stores the key instead of secret and does the signing with the key. We send the hash over REST API for example and get the signature back. The public key can be delivered to receiving party or the same REST API call can be made to the same keyvault. The keyvault can be backed by software or by real HSM, depending purely on the budget. Naturally, access policy and firewalling must be considered here.

Reply:
Nix doesn't support any of this yet, it insists on having the private key available as a file. That's the current state of things.

If you want to go into patching Nix, there's plans to change this, and a draft PR, but none of this is likely to land in the very short term.

Reply:
ok, since we are not using nix-serve, as discussed, we might as well just sign in Jenkins in "post build" mode. There is nothing that would prevent us from running it as 1 extra post-build step, right?

Reply:
Signing store paths happens on all builds, and is handled by the nix daemon, not Jenkins directly.

Jenkins doesn't have access to the signing keys or any way to influence the uploading process, and this ensures all builds are captured. But we can extend/change the post-build-hook script to do something else for signing, yes.